### PR TITLE
2022.9.x feature record revisions

### PR DESCRIPTION
**Pull request by darh**
Please take a look.

### CODE REVIEW

File: compose/rest/record.go [hunk 1]
```diff
@@ -567,6 +571,34 @@ func (ctrl *Record) TriggerScriptOnList(ctx context.Context, r *request.RecordTr
...
+		err = dal.IteratorEncodeJSON(ctx, w, iter, makeRev)
+		if err != nil {
+			return
+		}
+
+		if _, err = w.Write([]byte(`]}`)); err != nil {
+			return
+		}
+
+		return
```
Comment (tjerman) [inline compose/rest/record.go:598]: I feel like this bit could be moved to some rest util. function (especially the opening and closing bytes

File: compose/types/module.go [hunk 2]
```diff
@@ -26,6 +26,16 @@ type (
...
 
+		Revisions struct {
+			Enabled bool `json:"enabled"`
+
+			DAL struct {
+				// where will revisions be stored
+				// can contain variables like {{module}}, {{namespace}}
+				Ident string `json:"ident"`
+			} `json:"dal"`
+		} `json:"revisions"`
```
Comment (tjerman) [inline compose/types/module.go]: Should this be moved to its own type definition?

File: compose/types/module.go [hunk 3]
```diff
@@ -26,6 +26,16 @@ type (
...
+
 		SystemFieldEncoding SystemFieldEncoding `json:"systemFieldEncoding"`
```
Comment (tjerman) [inline compose/types/module.go]: In here, you'll also need to register the new revision system feld

Comment (tjerman) [reply compose/types/module.go]: In here as in the struct def.

File: compose/service/module.go [hunk 4]
```diff
@@ -1223,12 +1219,49 @@ func moduleToModel(ctx context.Context, dmm dalModelManager, ns *types.Namespace
...
+	if !ok {
+		err = fmt.Errorf("invalid model ident generated: %s", ident)
+		return
+	}
+
+	return ident, nil
+}
+
 // moduleFieldsToAttributes converts all user-defined module fields to attributes
 func moduleFieldsToAttributes(ctx context.Context, cm dal.ConnectionMeta, ns *types.Namespace, mod *types.Module) (out dal.AttributeSet, err error) {
```
Comment (tjerman) [inline compose/service/module.go:1282]: There is another fnc you'll need to update along with the struct change mentioned above (the added revision sys field comment) -- `partitionedModuleSystemFieldsToAttributes`

File: compose/service/record.go [hunk 5]
```diff
@@ -435,6 +442,43 @@ func (svc record) findSensitive(ctx context.Context, userID uint64, namespace *t
...
+		aProps.setNamespace(ns)
+		aProps.setRecord(rec)
+
+		if !mod.ModelConfig.Revisions.Enabled {
+			return RecordErrRevisionsDisabledOnModule()
+		}
+
+		if !svc.ac.CanSearchRevisionsOnRecord(ctx, rec) {
+			return RecordErrNotAllowedToSearchRevisions()
+		}
```
Comment (tjerman) [inline compose/service/record.go:473]: should we also have a `read` perm. for specific revisions?
I'd say no, but to consider

File: compose/service/record.go [hunk 6]
```diff
@@ -617,6 +661,13 @@ func (svc record) create(ctx context.Context, new *types.Record) (rec *types.Rec
 		return
 	}
 
+	// store revision
+	if m.ModelConfig.Revisions.Enabled {
+		if err = svc.revisions.created(ctx, new); err != nil {
+			return
+		}
+	}
```
Comment (tjerman) [inline compose/service/record.go]: We discussed doing revisions as a delta where the old value is logged and the record outlines the latest state.
What made it change?

File: provision/000_base/compose_access_control.yaml [hunk 7]
```diff
@@ -60,6 +60,8 @@ allow:
       - read
       - update
       - delete
+      - revisions.search
+
```
Comment (tjerman) [inline provision/000_base/compose_access_control.yaml:64]: rm extra new line

File: compose/service/record.go (hunk 5 above)
Comment (darh) [reply compose/service/record.go:473]: no

File: compose/service/record.go (hunk 6 above)
Comment (darh) [reply compose/service/record.go]: feels a bit less magical.

File: compose/rest/record.go (hunk 1 above)
Comment (darh) [reply compose/rest/record.go:598]: A agree. As soon as this proves stable an we have a 2nd place were we can use it

**Pull request merged**